rustls/src/verify.rs

```diff
@@ -457,60 +459,126 @@ impl UnparsedCertRevocationList {
+    ///
+    /// # Errors
+    /// This function will return [`RootsRequiredError`] if the [RootCertStore] is empty. If you
+    /// wish to disable client authentication use [WebPkiClientVerifier::no_client_auth()] instead.
+    pub fn builder(roots: RootCertStore) -> Result<ClientCertVerifierBuilder, RootsRequiredError> {
+        ClientCertVerifierBuilder::new(roots)
+    }
+
+    /// Create a new `WebPkiClientVerifier` that disables client authentication. The server will
+    /// not offer client authentication and anonymous clients will be accepted.
+    ///
+    /// This is in contrast to using `WebPkiClientVerifier::builder().allow_unauthenticated().build()`,
+    /// which will produce a verifier that will offer client authentication, but not require it.
+    pub fn no_client_auth() -> Arc<dyn ClientCertVerifier> {
+        Arc::new(NoClientAuth {})
+    }
+
+    /// Construct a new `WebpkiClientVerifier`.
     ///
     /// `roots` is the list of trust anchors to use for certificate validation.
-    pub fn new(roots: RootCertStore) -> Self {
+    /// `crls` are an iterable of owned certificate revocation lists (CRLs) to use for
+    /// client certificate validation.
+    /// `anonymous_policy` controls whether client authentication is required, or if anonymous
+    /// clients can connect.
+    pub(crate) fn new(
+        roots: RootCertStore,
+        crls: Vec<webpki::OwnedCertRevocationList>,
+        anonymous_policy: AnonymousClientPolicy,
+    ) -> Self {
         Self {
             subjects: roots
                 .roots
                 .iter()
                 .map(|r| r.subject().clone())
                 .collect(),
-            crls: Vec::new(),
+            crls,
             roots,
+            anonymous_policy,
         }
     }
-
-    /// Update the verifier to validate client certificates against the provided DER format
-    /// unparsed certificate revocation lists (CRLs).
-    pub fn with_crls(
-        self,
-        crls: impl IntoIterator<Item = UnparsedCertRevocationList>,
-    ) -> Result<Self, CertRevocationListError> {
-        Ok(Self {
-            crls: crls
-                .into_iter()
-                .map(|der_crl| der_crl.parse())
-                .collect::<Result<Vec<_>, CertRevocationListError>>()?,
-            ..self
-        })
-    }
-
-    /// Wrap this verifier in an [`Arc`] and coerce it to `dyn ClientCertVerifier`
-    #[inline(always)]
-    pub fn boxed(self) -> Arc<dyn ClientCertVerifier> {
-        // This function is needed because `ClientCertVerifier` is only reachable if the
-        // `dangerous_configuration` feature is enabled, which makes coercing hard to outside users
-        Arc::new(self)
-    }
 }
 
-impl ClientCertVerifier for AllowAnyAuthenticatedClient {
+impl ClientCertVerifier for WebPkiClientVerifier {
     fn offer_client_auth(&self) -> bool {
         true
     }
 
+    fn client_auth_mandatory(&self) -> bool {
```

Comment:
Since the API is changing anyway, should we change the trait method signature to use the `enum` instead of a bool (and, if so, maybe change the name to match)?

Reply:
I lightly prefer the trait as-is. I think it's easier to implement defaults at the trait level as-is.

Reply:
Fair enough.